bluray_player.1
'\" t
.\"     Title: bluray_player
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.23
.\"      Date: 2025-06-11
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "BLURAY_PLAYER" "1" "2025-06-11" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
bluray_player \- play a Blu\-ray using libmpv
.SH "SYNOPSIS"
.sp
\fBbluray_player\fP [\fIPATH\fP] [\fIOPTIONS\fP]
.SH "DESCRIPTION"
.sp
The bluray_player(1) program will play a playlist from a Blu\-ray disc,
image, or directory. Playback is handled by by libmpv.
.sp
Input path can be a single filename (image), a directory, or a device name.
The default device is based on your operating system, and is the primary
optical drive.
.SH "OPTIONS"
.sp
\fB\-m, \-\-main\fP
.RS 4
Play the main playlist (default). You can see which one is the main
one using \fIbluray_info \-\-main\fP. MPV filters out duplicate playlists,
so use \fIbluray_info \-\-no\-duplicates\fP as an index.
.RE
.sp
\fB\-p, \-\-playlist\fP \fIPLAYLIST\fP
.RS 4
Play the selected playlist number. MPV does not support selecting
playlist from non\-duplicate titles \(em when running bluray_info, so
use \fI\-\-no\-duplicates\fP to see which one you want to play.
.RE
.sp
\fB\-c, \-\-chapters\fP \fICHAPTER[\-[CHAPTER]]\fP
Play the selected chapter range. Default is to play all chapters of
the playlist.
.sp
\fB\-a, \-\-alang\fP \fILANGUAGE\fP
Set default audio language. Format is three\-letter language code, ISO
639\-2. For example: eng, deu, fra, spa.
.sp
\fB\-A, \-\-aid\fP \fISTREAM ID\fP
Play audio stream ID indexed by mpv.
.sp
\fB\-s, \-\-slang\fP \fILANGUAGE\fP
Display subtitles for selected language, if available.
.sp
\fB\-S, \-\-sid\fP \fISTREAM ID\fP
Play subtitle stream ID indexed by mpv.
.sp
\fB\-V, \-\-vid\fP \fISTREAM ID\fP
Play video stream ID indexed by mpv (default: first).
.sp
\fB\-f, \-\-fullscreen\fP
Display video in fullscreen.
.sp
\fB\-d, \-\-deinterlace\fP
Deinterlace video during playback.
.sp
\fB\-h, \-\-help\fP
Display help output.
.sp
\fB\-\-version\fP
Display version information.
.SH "CONFIGURATION"
.sp
bluray_player(1) uses libmpv as the engine for playback (see
.URL "https://mpv.io/" "" ")."
Configuration for bluray_player is stored in
~/.config/bluray_player/mpv.conf and will be read on each time during startup.
.sp
Passing options to \fIbluray_player\fP will override the configuration file.
.sp
See mpv(1) for all configuration options.
.SH "DECRYPTION"
.sp
The other bluray applications support setting a custom location for your
KEYDB.cfg file. bluray_player does not. It must be located in your home
directory at \fI.config/aacs/KEYDB.cfg\fP.
.SH "SEE ALSO"
.sp
bluray_info(1), bluray_copy(1), bluray_backup(1), bluray_rip(1), mpv(1)
.SH "BUGS"
.sp
\fBbluray_player\fP does not wait for an optical device to be ready. Wait for it to
finish "polling" before running the program.
.sp
Please file bugs at \c
.URL "https://github.com/beandog/bluray_info/issues" "" ""
.SH "AUTHOR"
.sp
bluray_player was written by Steve Dibb \c
.MTO "steve.dibb\(atgmail.com" "" ""
.SH "RESOURCES"
.sp
Source code available at GitHub: \c
.URL "https://github.com/beandog/bluray_info" "" ""
.sp
Main web site: \c
.URL "https://dvds.beandog.org" "" ""
.SH "COPYING"
.sp
Copyright (C) 2019 Steve Dibb. Free use of this software is granted under the
terms of the GNU General Public License, version 2 (GPL).

ChangeLog
ChangeLog

1.15
- Add bluray_backup
- Add bluray_rip - an ffmpeg wrapper using bluray_copy

1.14

- Remove option to set titles as input, use only playlists
- bluray_info: fix '--help' output
- bluray_info: display some metadata found in XML file
- bluray_info: choosing a language for subtitles or audio streams
  can be done for each type individually
- bluray_player: fix opening player fullscreen

1.13

- bluray_info: add options to filter by language, for either audio
  streams or subtitles

1.12

- bluray_info: add option to display 'duplicate' titles that would
  normally be filtered by libbluray. See bug #11
- bluray_copy: also add option to select from duplicate titles

1.11

- Fix configure script so it can build without libmpv (see
  Gentoo bug #907865)

1.10

bluray_info:
- Fix JSON output when limiting tracks (bug #13, thanks to zscholl
  for reporting)

bluray_copy:
- Display filename being written

1.9

- Fix not showing video format on 4K

1.8

- Make output more human-friendly
- Require libbluray >= 1.2.0 to fix 4K support
- Require libaacs

1.7

- Display milliseconds for timestamps and durations, (patch thanks to
  rapperskull)

1.6

- Display bluray chapter filesizes

1.5

- Add bluray_player, a small Blu-ray player using libmpv
- Add man pages
- Add support for 4K

bluray_info:

- Properly label DTS-HD Master audio streams
- Add displaying long audio and video codec names in JSON output
- Can narrow results on minimum number of seconds or minutes, has an audio or
  subtitle track
- Display chapter start times
- Can output chapters in OGM format for muxing into MKV
- Set disc title from Blu-ray disc metadata instead of UDF volume name

bluray_copy:

- General rewrite for readability and stability
- Fixed pulling in extra data between title start and first chapter
- Check for AACS decrypt errors
- Copy angle numbers

1.4

- Fix indexing to match Blu-ray spec for playlist numbers
- bluray_info: display main title even if one track is listed
- bluray_info: display subtitle character code
- bluray_copy: can now stream to stdout
- bluray_copy: print errors to stderr
- bluray_copy: increase default buffer sizes
- bluray_copy: increased buffer size for Blu-ray optical drives
- bluray_copy: don't require an output filename, use defaults if none given
- bluray_copy: optional quiet output to remove printing copy progress

1.3

- Add autotools

1.2.0
- initialize structs properly
- bluray_copy quits on disk space full

1.1.0
- add bluray_copy

1.0.0
- Initial release


bluray_player.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <string.h>
#include <stdbool.h>
#include <inttypes.h>
#include <math.h>
#include <getopt.h>
#include <sys/stat.h>
#include <fcntl.h>
#ifdef __linux__
#include <linux/limits.h>
#else
#include <limits.h>
#endif
#include <libbluray/bluray.h>
#include "bluray_device.h"
#include "bluray_open.h"
#include "bluray_time.h"
#include "bluray_player.h"
#include <mpv/client.h>

/**
 *   _     _                                  _
 *  | |__ | |_   _ _ __ __ _ _   _      _ __ | | __ _ _   _  ___ _ __
 *  | '_ \| | | | | '__/ _` | | | |    | '_ \| |/ _` | | | |/ _ \ '__|
 *  | |_) | | |_| | | | (_| | |_| |    | |_) | | (_| | |_| |  __/ |
 *  |_.__/|_|\__,_|_|  \__,_|\__, |____| .__/|_|\__,_|\__, |\___|_|
 *                           |___/_____|_|            |___/
 *
 * bluray video player using libbluray and libmpv
 *
 * If your Blu-ray is encrypted, you'll need libaacs as well as a modern
 * KEYDB.cfg file to access them.
 *
 * More information at https://dvds.beandog.org/
 *
 * Go watch tt0067992 :)
 *
 */

int main(int argc, char **argv) {

	char device_filename[PATH_MAX] = {'\0'};
	bool opt_playlist_number = false;
	bool opt_chapter_start = false;
	bool opt_chapter_end = false;
	bool exit_help = false;
	bool opt_video_stream = false;
	bool opt_audio_stream = false;
	bool opt_subtitle_stream = false;
	char stream_id[4];
	unsigned long int arg_number = 0;
	uint32_t arg_playlist_number = 0;
	uint32_t arg_first_chapter = 0;
	uint32_t arg_last_chapter = 0;
	const char *home_dir = getenv("HOME");
	int retval = 0;

	struct bluray_player bluray_player;
	snprintf(bluray_player.config_dir, BLURAY_PLAYER_CONFIG_DIR_STRLEN, "/.config/bluray_player");
	memset(bluray_player.mpv_config_dir, '\0', sizeof(bluray_player.mpv_config_dir));
	if(home_dir != NULL)
		snprintf(bluray_player.mpv_config_dir, BLURAY_PLAYER_PATH_MAX, "%s%s", home_dir, bluray_player.config_dir);

	struct bluray_playback bluray_playback;
	bluray_playback.title = 0;
	bluray_playback.fullscreen = false;
	bluray_playback.detelecine = false;
	memset(bluray_playback.audio_lang, '\0', BLURAY_PLAYER_LANG_STRLEN);
	memset(bluray_playback.subtitles_lang, '\0', BLURAY_PLAYER_LANG_STRLEN);
	memset(bluray_playback.chapter_start, '\0', BLURAY_PLAYER_CHAPTER_STRLEN);
	memset(bluray_playback.chapter_end, '\0', BLURAY_PLAYER_CHAPTER_STRLEN);

	char *token = NULL;
	int g_opt = 0;
	int g_ix = 0;
	struct option p_long_opts[] = {
		{ "alang", required_argument, NULL, 'a' },
		{ "aid", required_argument, NULL, 'A' },
		{ "chapters", required_argument, NULL, 'c' },
		{ "detelecine", no_argument, NULL, 'd' },
		{ "fullscreen", no_argument, NULL, 'f' },
		{ "help", no_argument, NULL, 'h' },
		{ "main", no_argument, NULL, 'm' },
		{ "playlist", required_argument, NULL, 'p' },
		{ "slang", required_argument, NULL, 's' },
		{ "sid", required_argument, NULL, 'S' },
		{ "vid", required_argument, NULL, 'V' },
		{ "version", no_argument, NULL, 'Z' },
		{ 0, 0, 0, 0 }
	};
	while((g_opt = getopt_long(argc, argv, "a:A:c:dfhmp:s:S:V:Z", p_long_opts, &g_ix)) != -1) {

		switch(g_opt) {

			case 'a':
				strncpy(bluray_playback.audio_lang, optarg, BLURAY_PLAYER_LANG_STRLEN - 1);
				break;

			case 'A':
				opt_audio_stream = true;
				arg_number = strtoul(optarg, NULL, 10);
				bluray_playback.audio_stream_id = (uint8_t)arg_number;
				break;

			case 'c':

				// passed -c -
				if(strlen(optarg) == 1 && strncmp(optarg, "-", 1) == 0) {
					printf("invalid syntax: - only, erroring out\n");
					return 1;
				}

				// passed -c <chapter number>
				if(strchr(optarg, '-') == NULL) {
					opt_chapter_start = true;
					opt_chapter_end = false;
					arg_number = strtoul(optarg, NULL, 10);
					if(arg_number < 2)
						arg_first_chapter = 1;
					else
						arg_first_chapter = (uint8_t)arg_number;
					break;
				}

				// passed -c -<chapter number>
				if(strncmp(optarg, "-", 1) == 0) {
					opt_chapter_start = false;
					opt_chapter_end = true;
					token = strtok(optarg, "-");
					arg_number = strtoul(token, NULL, 10);
					if(arg_number < 2)
						arg_last_chapter = 1;
					else
						arg_last_chapter = (uint8_t)arg_number;
					break;
				}

				// passed -c <chapter-number>-[|<chapter-number>]
				opt_chapter_start = true;
				token = strtok(optarg, "-");
				arg_number = strtoul(token, NULL, 10);
				if(arg_number < 2)
					arg_first_chapter = 1;
				else
					arg_first_chapter = (uint8_t)arg_number;

				token = strtok(NULL, "-");

				// passed -c <chapter-number>-<chapter-number>
				if(token != NULL) {
					opt_chapter_end = true;
					arg_number = strtoul(token, NULL, 10);
					if(arg_number < 2)
						arg_last_chapter = 1;
					else
						arg_last_chapter = (uint8_t)arg_number;
					break;
				}

				// at this point, only passed -c <chapter-number>- which is
				// handled the same as -c <chapter-number>
				opt_chapter_end = false;

				break;

			case 'd':
				bluray_playback.detelecine = true;
				break;

			case 'f':
				bluray_playback.fullscreen = true;
				break;

			case 'm':
				opt_playlist_number = false;
				break;

			case 'p':
				opt_playlist_number = true;
				arg_number = strtoul(optarg, NULL, 10);
				arg_playlist_number = (uint32_t)arg_number;
				break;

			case 's':
				strncpy(bluray_playback.subtitles_lang, optarg, BLURAY_PLAYER_LANG_STRLEN - 1);
				break;

			case 'S':
				opt_subtitle_stream = true;
				arg_number = strtoul(optarg, NULL, 10);
				bluray_playback.subtitle_stream_id = (uint8_t)arg_number;
				break;

			case 'V':
				opt_video_stream = true;
				arg_number = strtoul(optarg, NULL, 10);
				bluray_playback.video_stream_id = (uint8_t)arg_number;
				break;

			case 'Z':
				printf("bluray_player " PACKAGE_VERSION "\n");
				exit_help = true;
				break;

			case 'h':
				printf("bluray_player - play a Blu-ray using libmpv\n");
				printf("\n");
				printf("Usage: bluray_player [bluray path] [options]\n");
				printf("\n");
				printf("Title selection:\n");
				printf("  -m, --main               Play main playlist (default)\n");
				printf("  -p, --playlist <#>       Play playlist number\n");
				printf("  -c, --chapters <#>[-#]   Play chapter number(s)\n");
				printf("\n");
				printf("Languages - ISO 639-2 three-letter language codes (eng, deu, fra, spa, ...):\n");
				printf("  -a, --alang <language>   Audio language (default: first stream)\n");
				printf("  -s, --slang <language>   Subtitles language (default: first stream)\n");
				printf("\n");
				printf("Stream IDs:\n");
				printf("  -V, --vid <number>       Play video stream id\n");
				printf("  -A, --aid <number>       Play audio stream id\n");
				printf("  -S, --sid <number>       Play subtitle stream id\n");
				printf("\n");
				printf("Playback:\n");
				printf("  -f, --fullscreen         Display fullscreen\n");
				printf("  -d, --detelecine         Detelecine video\n");
				printf("\n");
				printf("Other:\n");
				printf("  -h, --help               This output\n");
				printf("      --version	           Display version\n");
				printf("\n");
				printf("MPV configuration is read from ~/.config/bluray_player/mpv.conf\n");
				printf("\n");
				printf("MPV ignores duplicate titles by default, so use 'bluray_info --no-duplicates' to get same playlist number.\n");
				printf("\n");
				printf("Blu-ray path can be a device, filename, or directory (default: " DEFAULT_BLURAY_DEVICE ")\n");
				exit_help = true;
				break;

			case 0:
			default:
				exit_help = true;
				break;

		}

	}

	if(exit_help)
		return 0;

	memset(device_filename, '\0', PATH_MAX);
	if (argv[optind])
		strncpy(device_filename, argv[optind], PATH_MAX - 1);
	else
		strncpy(device_filename, DEFAULT_BLURAY_DEVICE, PATH_MAX - 1);

	// Open device - MPV does *not* support opening custom location for
	// KEYDB file, so default location will have to work. (0.37.0)
	BLURAY *bd = NULL;
	bd = bd_open(device_filename, NULL);

	if(bd == NULL) {
		fprintf(stderr, "Could not open Blu-ray %s\n", device_filename);
		return 1;
	}

	// Fetch info
	const BLURAY_DISC_INFO *bd_info = NULL;

	bd_info = bd_get_disc_info(bd);

	if(bd_info == NULL) {
		fprintf(stderr, "Could not get Blu-ray disc info\n");
		return 1;
	}

	// Blu-ray
	struct bluray_info bluray_info;
	// MPV does not list duplicate titles, so ignore them
	retval = bluray_info_init(bd, &bluray_info, false);

	struct bluray_title bluray_title;

	// Select title passed as an argument
	if(opt_playlist_number) {
		retval = bd_select_playlist(bd, arg_playlist_number);
		if(retval == 0) {
			fprintf(stderr, "Could not open playlist %" PRIu32 "\n", arg_playlist_number);
			bd_close(bd);
			bd = NULL;
			return 1;
		}
		bluray_title.ix = bd_get_current_title(bd);
		bluray_title.playlist = arg_playlist_number;
	} else {
		bluray_title.ix = bluray_info.main_title;
		retval = bd_select_title(bd, bluray_title.ix);
		if(retval == 0) {
			fprintf(stderr, "Could not open main playlist");
			bd_close(bd);
			bd = NULL;
			return 1;
		}
	}

	// Init bluray_title struct
	uint8_t angle_ix = 0;
	// We are filtering out duplicate titles here, because that is what mpv does,
	// current version is at 0.37.0.
	retval = bluray_title_init(bd, &bluray_title, bluray_title.ix, angle_ix, false);

	if(retval) {
		fprintf(stderr, "Could not open playlist %" PRIu32 ", quitting\n", bluray_title.playlist);
		return 1;
	}

	// Silently check and fix chapter boundaries for playback
	if(arg_last_chapter > 0 && arg_last_chapter > bluray_title.chapters)
		arg_last_chapter = bluray_title.chapters;
	if(arg_last_chapter > 0 && arg_first_chapter > arg_last_chapter)
		arg_first_chapter = arg_last_chapter;

	// When choosing a chapter range, mpv will add 1 to the last one requested
	// For example, to play just chapter 3, send playback of 3-4
	memset(bluray_playback.chapter_start, '\0', BLURAY_PLAYER_CHAPTER_STRLEN);
	snprintf(bluray_playback.chapter_start, BLURAY_PLAYER_CHAPTER_STRLEN, "#%" PRIu32, arg_first_chapter);
	memset(bluray_playback.chapter_end, '\0', BLURAY_PLAYER_CHAPTER_STRLEN);
	snprintf(bluray_playback.chapter_end, BLURAY_PLAYER_CHAPTER_STRLEN, "#%" PRIu32, arg_last_chapter + 1);

	// MPV zero-indexes title numbers
	bluray_playback.title = bluray_title.ix;

	if(strlen(bluray_info.disc_name))
		printf("[bluray_player] %s\n", bluray_info.disc_name);

	// libmpv doesn't support Blu-ray angle selection (as of latest stable, 0.29.1)
	printf("[bluray_player] Playlist: %*" PRIu32 ", Length: %s, Chapters: %*" PRIu32 ", Video streams: %*" PRIu8 ", Audio streams: %*" PRIu8 ", Subtitles: %*" PRIu8 ", Angles: %*" PRIu8 ", Filesize: %*" PRIu64 " MBs\n", 5, bluray_title.playlist, bluray_title.length, 2, bluray_title.chapters, 2, bluray_title.video_streams, 2, bluray_title.audio_streams, 2, bluray_title.pg_streams, 2, bluray_title.angles, 5, bluray_title.size_mbs);

	// Finished with libbluray
	bd_close(bd);
	bd = NULL;

	// Note that the order and location of setting mpv configuration is important,
	// especially if you want to override mpv.conf in ~/.config/bluray_player/

	// Blu-ray playback using libmpv
	mpv_handle *bluray_mpv = NULL;
	bluray_mpv = mpv_create();
	if(bluray_mpv == NULL) {
		fprintf(stderr, "[bluray_player] Could not create an MPV instance, quitting\n");
		return 1;
	}

	// Load user's mpv configuration in ~/.config/bluray_player/mpv.conf (and friends)
	if(strlen(bluray_player.mpv_config_dir) > 0) {
		mpv_set_option_string(bluray_mpv, "config-dir", bluray_player.mpv_config_dir);
		mpv_set_option_string(bluray_mpv, "config", "yes");
	}

	// MPV doesn't need much before initializing, and config-dir is the only relevant one here
	retval = mpv_initialize(bluray_mpv);
	if(retval) {
		fprintf(stderr, "[bluray_player] Could not initalize MPV, quiting\n");
		return 1;
	}

	// Playback options and default configuration
	mpv_set_option_string(bluray_mpv, "bluray-device", device_filename);
	mpv_set_option_string(bluray_mpv, "terminal", "yes");
	mpv_set_option_string(bluray_mpv, "term-osd-bar", "yes");
	mpv_set_option_string(bluray_mpv, "input-default-bindings", "yes");
	mpv_set_option_string(bluray_mpv, "input-vo-keyboard", "yes");
	if(bluray_playback.fullscreen)
		mpv_set_option_string(bluray_mpv, "fullscreen", "yes");
	if(bluray_playback.detelecine)
		mpv_set_option_string(bluray_mpv, "vf", "pullup,dejudder");
	if(opt_chapter_start && arg_first_chapter > 0)
		mpv_set_option_string(bluray_mpv, "start", bluray_playback.chapter_start);
	if(opt_chapter_end && arg_last_chapter > 0)
		mpv_set_option_string(bluray_mpv, "end", bluray_playback.chapter_end);

	// mpv zero-indexes titles
	char bluray_mpv_args[BLURAY_PLAYER_TITLE_STRLEN];
	memset(bluray_mpv_args, '\0', BLURAY_PLAYER_TITLE_STRLEN);
	snprintf(bluray_mpv_args, BLURAY_PLAYER_TITLE_STRLEN, "bd://%" PRIu32, bluray_playback.title);

	// MPV uses zero-indexing for titles, bluray_info uses one instead
	const char *bluray_mpv_commands[] = {
		"loadfile",
		bluray_mpv_args,
		NULL
	};

	retval = mpv_command(bluray_mpv, bluray_mpv_commands);
	if(retval) {
		fprintf(stderr, "[bluray_player] Could not send MPV arguments: %s; using defaults\n", bluray_mpv_args);
	}

	// Based on the decoder used in libmpv, it may or may not find other streams.
	// bluray_player intentionally doesn't check for out of bounds or non-existing
	// audio and subtitle languages and tracks, and lets the player handle user input.

	// Any arguments after this part will override mpv.conf

	// Set playback languages in order of language code, then stream IDs
	// When selecting a lanugage with --alang or --slang, it will choose the first of
	// any streams with that language. Setting --aid, or --sid will choose the specific one.
	if(strlen(bluray_playback.audio_lang) > 0)
		retval = mpv_set_option_string(bluray_mpv, "alang", bluray_playback.audio_lang);
	if(strlen(bluray_playback.subtitles_lang) > 0)
		mpv_set_option_string(bluray_mpv, "slang", bluray_playback.subtitles_lang);

	// I vaguely recall seeing a Blu-ray with two video streams before
	if(opt_video_stream) {
		memset(stream_id, '\0', 4);
		snprintf(stream_id, 4, "%" PRIu8, bluray_playback.video_stream_id);
		retval = mpv_set_option_string(bluray_mpv, "vid", stream_id);
	}
	if(opt_audio_stream) {
		memset(stream_id, '\0', 4);
		snprintf(stream_id, 4, "%" PRIu8, bluray_playback.audio_stream_id);
		retval = mpv_set_option_string(bluray_mpv, "aid", stream_id);
	}
	if(opt_subtitle_stream) {
		memset(stream_id, '\0', 4);
		snprintf(stream_id, 4, "%" PRIu8, bluray_playback.subtitle_stream_id);
		retval = mpv_set_option_string(bluray_mpv, "sid", stream_id);
	}

	mpv_event *bluray_mpv_event = NULL;
	struct mpv_event_log_message *bluray_mpv_log_message = NULL;

	while(true) {

		bluray_mpv_event = mpv_wait_event(bluray_mpv, -1);

		if(bluray_mpv_event->event_id == MPV_EVENT_SHUTDOWN || bluray_mpv_event->event_id == MPV_EVENT_END_FILE)
			break;

		if(bluray_mpv_event->event_id == MPV_EVENT_LOG_MESSAGE) {
			bluray_mpv_log_message = (struct mpv_event_log_message *)bluray_mpv_event->data;
			fprintf(stderr, "[bluray_player] mpv [%s]: %s", bluray_mpv_log_message->level, bluray_mpv_log_message->text);
		}

	}

	mpv_terminate_destroy(bluray_mpv);

	return 0;

}
